File: src/pages/component/Trading/CloseMarket.py
import json
import csv
import datetime

# Dictionary to map projectType strings to numbers
projectTypeDict = {
    "Renewable Energy": 1,
    "Energy Efficiency": 2,
    "Forestry and Land Use": 3,
    "Agricultural Land Management": 4,
    "Waste Management": 5,
    "Industrial Processes": 6,
    "Transportation": 7,
    "Blue Carbon": 8
}

# File paths
json_file_path = r"C:\Users\ASUS\OneDrive - Monash University\Back ups\Desktop\FYP Webpage\carbontrading\src\pages\component\Data\BidsMade.json"
csv_file_path =  r"C:\Users\ASUS\OneDrive - Monash University\FYP\MyEthereumProject\Data\datav2.csv"
user_id = "SFqZK5ev"

def load_bidding_data():
    try:
        with open(json_file_path, "r", encoding="utf-8") as file:
            data = json.load(file)
        return data.get("bids", [])
    except Exception as e:
        print("Error reading JSON file:", e)
        return []

def append_to_csv(data):
    file_exists = False
    try:
        with open(csv_file_path, "r", encoding="utf-8") as file:
            file_exists = True
    except FileNotFoundError:
        pass
    
    with open(csv_file_path, "a") as file:
        writer = csv.writer(file)
        
        if not file_exists:
            writer.writerow(["Timestamp", "User ID", "Type", "Carbon Units", "Unit Price", "Total Price", "Project Type"])
        
        newdata = []
        for bid in data:
            timestamp = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            project_type = projectTypeDict.get(bid["projectType"], "Unknown")
            total_price = bid["CarbonCredits"] * bid["unitPrice"]
            
            newdata.append([timestamp, user_id, bid["bidType"], bid["CarbonCredits"], bid["unitPrice"], total_price, project_type])
        # print(newdata)
        writer.writerows(newdata)
    print("Bidding data appended successfully!")

def clear_bidding_data():
    with open(json_file_path, "w") as file:
        json.dump({"bids": []}, file)
    print("Bidding data cleared successfully!")

if __name__ == "__main__":
    bidding_data = load_bidding_data()
    append_to_csv(bidding_data)
    clear_bidding_data()

File: src/pages/component/Trading/CloseMarket.js
const fs = require("fs");
const path = require("path");
const { parse: parseCsv } = require("csv-parse");

const projectTypeDict = {
    "Renewable Energy": 1,
    "Energy Efficiency": 2,
    "Forestry and Land Use": 3,
    "Agricultural Land Management": 4,
    "Waste Management": 5,
    "Industrial Processes": 6,
    "Transportation": 7,
    "Blue Carbon": 8
};

const userID = "SFqZK5ev";
const csvFilePath = path.join(__dirname, "../Data/datav2.csv");
const jsonFilePath = path.join(__dirname, "../Data/BidsMade.json");

function loadBiddingData() {
    try {
        const jsonData = JSON.parse(fs.readFileSync(jsonFilePath, "utf8"));
        return jsonData.bids || [];
    } catch (error) {
        console.error("Error reading JSON file:", error);
        return [];
    }
}

function appendToCsv(data) {
    // append json data to csv file
    const csvStream = fs.createReadStream(csvFilePath).pipe(parseCsv({ columns: true, relax_column_count: true }));
    const newRows = [];
    csvStream.on("data", (row) => {
        const newRow = { ...row };
        if (newRow.userID === userID) {
            const bid = data.find((d) => d.projectID === newRow.projectID);
            if (bid) {
                newRow.biddingPrice = bid.biddingPrice;
            }
        }
        newRows.push(newRow);
    });
    csvStream.on("end", () => {
        const newCsvData = newRows.map((row) => {
            return `${row.projectID},${row.projectName},${row.projectType},${row.projectOwner},${row.biddingPrice}`;
        });
        fs.writeFileSync(csvFilePath, newCsvData.join("\n"));
        console.log("Bidding data appended successfully!");
    });
    // console.log("Bidding data appended successfully!");
}

const biddingData = loadBiddingData();
appendToCsv(biddingData);
